### PR TITLE
refactor(llmq): separate llmq/utils into llmq/utils and llmq/complex_…

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
…utils

llmq/utils should remain very slim including only trivial headers, while llmq/complex_utils for the time being may continue to include more complex includes such as
```
#include <llmq/quorums.h>
#include <llmq/commitment.h>
#include <llmq/snapshot.h>
#include <evo/deterministicmns.h>
#include <masternode/meta.h>
```

This helps us break some circular dependancies and makes llmq/utils a nice simple neighborhood to live. This also includes substancial include refactoring as well as moving a function into its own util files

note that the circular depends linter had a lot of exclusions added, but I think that's an artifact of the fact that c(omplex) is earlier in the alphabet than u(tils)

### CODE REVIEW

**Comment by UdjinM6** [comment]
linter complains

File: src/util/hexstr.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+#ifndef BITCOIN_UTIL_GETUNIQUEPATH_H
+#define BITCOIN_UTIL_GETUNIQUEPATH_H
+
+#include <vector>
+#include <string>
+
+std::string BitsToHexStr(const std::vector<bool>& vBits);
+
+#endif // BITCOIN_UTIL_GETUNIQUEPATH_H
```
Comment (UdjinM6) [inline src/util/hexstr.h:13]:
```suggestion
#endif // BITCOIN_UTIL_GETUNIQUEPATH_H

```

File: test/lint/lint-circular-dependencies.sh [hunk 2]
```diff
@@ -109,7 +124,7 @@ for CIRC in $(cd src && ../contrib/devtools/circular-dependencies.py {*,*/*,*/*/
         fi
     done
     if [[ ${IS_EXPECTED_CIRC} == 0 ]]; then
-        echo "A new circular dependency in the form of \"${CIRC}\" appears to have been introduced."
+        echo "\"${CIRC}\""
```
Comment (UdjinM6) [inline test/lint/lint-circular-dependencies.sh:127]:
```suggestion
        echo "A new circular dependency in the form of \"${CIRC}\" appears to have been introduced."
```

File: src/util/hexstr.cpp [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) 2022 The Dash Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <util/strencodings.h>
+
+#include <string>
+#include <vector>
+
+std::string BitsToHexStr(const std::vector<bool>& vBits)
```
Comment (UdjinM6) [inline src/util/hexstr.cpp:10]: why not in `util/strencodings.*`?

File: src/llmq/commitment.cpp [hunk 4]
```diff
@@ -159,6 +161,23 @@ bool CFinalCommitment::VerifySizes(const Consensus::LLMQParams& params) const
...
+//    obj.pushKV("signersCount", CountSigners());
+//    obj.pushKV("signers", CLLMQUtils::ToHexStr(signers));
+//    obj.pushKV("validMembersCount", CountValidMembers());
+//    obj.pushKV("validMembers", CLLMQUtils::ToHexStr(validMembers));
+//    obj.pushKV("quorumPublicKey", quorumPublicKey.ToString());
+//    obj.pushKV("quorumVvecHash", quorumVvecHash.ToString());
+//    obj.pushKV("quorumSig", quorumSig.ToString());
+//    obj.pushKV("membersSig", membersSig.ToString());
+//}
+//
```
Comment (UdjinM6) [inline src/llmq/commitment.cpp:180]: ?

File: src/llmq/commitment.cpp [hunk 5]
```diff
@@ -216,4 +235,15 @@ bool CheckLLMQCommitment(const CTransaction& tx, const CBlockIndex* pindexPrev,
...
+//{
+//    obj.setObject();
+//    obj.pushKV("version", int(nVersion));
+//    obj.pushKV("height", int(nHeight));
+//
+//    UniValue qcObj;
+//    commitment.ToJson(qcObj);
+//    obj.pushKV("commitment", qcObj);
+//}
+
```
Comment (UdjinM6) [inline src/llmq/commitment.cpp:248]: ?

File: src/llmq/commitment.h [hunk 6]
```diff
@@ -127,6 +132,8 @@ class CFinalCommitmentTxPayload
     CFinalCommitment commitment;
 
 public:
+//    void ToJson(UniValue& obj) const;
+
```
Comment (UdjinM6) [inline src/llmq/commitment.h:136]: ?

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Review by ogabrielides — CHANGES_REQUESTED** [review]
@PastaPastaPasta Please apply those changes (I have no access to push)

```
diff --git a/src/llmq/commitment.cpp b/src/llmq/commitment.cpp
index 6b436dd2f9..8e2ca96490 100644
--- a/src/llmq/commitment.cpp
+++ b/src/llmq/commitment.cpp
@@ -161,23 +161,6 @@ bool CFinalCommitment::VerifySizes(const Consensus::LLMQParams& params) const
     return true;
 }
 
-//void CFinalCommitment::ToJson(UniValue& obj) const
-//{
-//    obj.setObject();
-//    obj.pushKV("version", (int)nVersion);
-//    obj.pushKV("llmqType", (int)llmqType);
-//    obj.pushKV("quorumHash", quorumHash.ToString());
-//    obj.pushKV("quorumIndex", quorumIndex);
-//    obj.pushKV("signersCount", CountSigners());
-//    obj.pushKV("signers", CLLMQUtils::ToHexStr(signers));
-//    obj.pushKV("validMembersCount", CountValidMembers());
-//    obj.pushKV("validMembers", CLLMQUtils::ToHexStr(validMembers));
-//    obj.pushKV("quorumPublicKey", quorumPublicKey.ToString());
-//    obj.pushKV("quorumVvecHash", quorumVvecHash.ToString());
-//    obj.pushKV("quorumSig", quorumSig.ToString());
-//    obj.pushKV("membersSig", membersSig.ToString());
-//}
-//
 bool CheckLLMQCommitment(const CTransaction& tx, const CBlockIndex* pindexPrev, CValidationState& state)
 {
     CFinalCommitmentTxPayload qcTx;
@@ -235,15 +218,5 @@ bool CheckLLMQCommitment(const CTransaction& tx, const CBlockIndex* pindexPrev,
     return true;
 }
 
-//void CFinalCommitmentTxPayload::ToJson(UniValue& obj) const
-//{
-//    obj.setObject();
-//    obj.pushKV("version", int(nVersion));
-//    obj.pushKV("height", int(nHeight));
-//
-//    UniValue qcObj;
-//    commitment.ToJson(qcObj);
-//    obj.pushKV("commitment", qcObj);
-//}
 
 } // namespace llmq
diff --git a/src/llmq/commitment.h b/src/llmq/commitment.h
index 9bbb03c8fd..24795bd1e0 100644
--- a/src/llmq/commitment.h
+++ b/src/llmq/commitment.h
@@ -132,7 +132,6 @@ public:
     CFinalCommitment commitment;
 
 public:
-//    void ToJson(UniValue& obj) const;
 
     SERIALIZE_METHODS(CFinalCommitmentTxPayload, obj)
     {
diff --git a/src/llmq/instantsend.cpp b/src/llmq/instantsend.cpp
index c6827f4c20..b05d99395a 100644
--- a/src/llmq/instantsend.cpp
+++ b/src/llmq/instantsend.cpp
@@ -8,6 +8,7 @@
 #include <llmq/quorums.h>
 #include <llmq/utils.h>
 #include <llmq/commitment.h>
+#include <llmq/complex_utils.h>
 
 #include <bls/bls_batchverifier.h>
 #include <chainparams.h>
@@ -508,7 +509,7 @@ void CInstantSendManager::ProcessTx(const CTransaction& tx, bool fRetroactive, c
     // block after we retroactively locked all transactions.
     if (!IsInstantSendMempoolSigningEnabled() && !fRetroactive) return;
 
-    if (!TrySignInputLocks(tx, fRetroactive, CLLMQUtils::GetInstantSendLLMQType(WITH_LOCK(cs_main, return ::ChainActive().Tip())), params)) {
+    if (!TrySignInputLocks(tx, fRetroactive, CLLMQComplexUtils::GetInstantSendLLMQType(WITH_LOCK(cs_main, return ::ChainActive().Tip())), params)) {
         return;
     }
 
@@ -682,7 +683,7 @@ void CInstantSendManager::HandleNewInputLockRecoveredSig(const CRecoveredSig& re
 
 void CInstantSendManager::TrySignInstantSendLock(const CTransaction& tx)
 {
-    const auto llmqType = CLLMQUtils::GetInstantSendLLMQType(WITH_LOCK(cs_main, return ::ChainActive().Tip()));
+    const auto llmqType = CLLMQComplexUtils::GetInstantSendLLMQType(WITH_LOCK(cs_main, return ::ChainActive().Tip()));
 
     for (auto& in : tx.vin) {
         auto id = ::SerializeHash(std::make_pair(INPUTLOCK_REQUESTID_PREFIX, in.prevout));
@@ -851,7 +852,7 @@ bool CInstantSendManager::PreVerifyInstantSendLock(const llmq::CInstantSendLock&
 bool CInstantSendManager::ProcessPendingInstantSendLocks()
 {
     const CBlockIndex* pBlockIndexTip = WITH_LOCK(cs_main, return ::ChainActive().Tip());
-    if (pBlockIndexTip && CLLMQUtils::GetInstantSendLLMQType(pBlockIndexTip) == Params().GetConsensus().llmqTypeDIP0024InstantSend) {
+    if (pBlockIndexTip && CLLMQComplexUtils::GetInstantSendLLMQType(pBlockIndexTip) == Params().GetConsensus().llmqTypeDIP0024InstantSend) {
         // Don't short circuit. Try to process deterministic and not deterministic islocks
         return ProcessPendingInstantSendLocks(true) & ProcessPendingInstantSendLocks(false);
     } else {
diff --git a/src/llmq/snapshot.cpp b/src/llmq/snapshot.cpp
index 7ab267b429..864bacbc8e 100644
--- a/src/llmq/snapshot.cpp
+++ b/src/llmq/snapshot.cpp
@@ -11,6 +11,7 @@
 #include <llmq/blockprocessor.h>
 #include <llmq/commitment.h>
 #include <llmq/utils.h>
+#include <llmq/complex_utils.h>
 
 #include <base58.h>
 #include <chainparams.h>
@@ -162,7 +163,7 @@ bool BuildQuorumRotationInfo(const CGetQuorumRotationInfo& request, CQuorumRotat
     }
 
     //Quorum rotation is enabled only for InstantSend atm.
-    Consensus::LLMQType llmqType = CLLMQUtils::GetInstantSendLLMQType(blockIndex);
+    Consensus::LLMQType llmqType = CLLMQComplexUtils::GetInstantSendLLMQType(blockIndex);
 
     // Since the returned quorums are in reversed order, the most recent one is at index 0
     const Consensus::LLMQParams& llmqParams = GetLLMQParams(llmqType);
diff --git a/src/rpc/rpcquorums.cpp b/src/rpc/rpcquorums.cpp
index ebe9d0b8a1..e2c1b90a9b 100644
--- a/src/rpc/rpcquorums.cpp
+++ b/src/rpc/rpcquorums.cpp
@@ -870,7 +870,7 @@ static UniValue verifyislock(const JSONRPCRequest& request)
             pBlockIndex = ::ChainActive()[signHeight];
         }
     }
-    auto llmqType = llmq::CLLMQUtils::GetInstantSendLLMQType(pBlockIndex);
+    auto llmqType = llmq::CLLMQComplexUtils::GetInstantSendLLMQType(pBlockIndex);
     // First check against the current active set, if it fails check against the last active set
     int signOffset{llmq::GetLLMQParams(llmqType).dkgInterval};
     return llmq::quorumSigningManager->VerifyRecoveredSig(llmqType, signHeight, id, txid, sig, 0) ||
-- 
2.32.0 (Apple Git-132)
```

**Comment by knst** [comment]
superseeded by #5753 and #5790
`llmq/utils.h` has now only quorum calculations; all "light" thinks moved out such as llmq/options.h

**Pull request closed without merging**